File: apps/android/KachowV8/app/src/main/java/com/kartik/mcqueencontroller/Protocol.kt
package com.kartik.mcqueencontroller

import java.util.Locale

object Protocol {
    const val DEFAULT_PORT = 5007
    const val TOKEN = "rc-car"
    const val MAX_STEERING = 1000
    const val MAX_THROTTLE_COMMAND = 1000

    fun newSessionId(): String = String.format(Locale.US, "%08x", System.nanoTime().toInt())

    fun hello(session: String, sequence: Long, timestampMs: Long): String =
        "H,$TOKEN,$session,$sequence,$timestampMs\n"

    fun command(
        session: String,
        sequence: Long,
        timestampMs: Long,
        steering: Int,
        throttle: Int,
        motorEnabled: Boolean,
    ): String =
        "C,$TOKEN,$session,$sequence,$timestampMs,$steering,$throttle,${if (motorEnabled) 1 else 0}\n"

    fun emergency(session: String, sequence: Long, timestampMs: Long): String =
        "E,$TOKEN,$session,$sequence,$timestampMs\n"

    fun parseStatus(raw: String): StatusPacket? {
        val fields = raw.trim().split(',')
        if (fields.size != 11 || fields[0] != "S") return null

        return runCatching {
            StatusPacket(
                session = fields[1],
                acknowledgedSequence = fields[2].toLong(),
                echoTimestampMs = fields[3].toLong(),
                commandedSteering = fields[4].toInt(),
                actualServoAngle = fields[5].toInt(),
                commandedThrottle = fields[6].toInt(),
                actualThrottle = fields[7].toInt(),
                failsafe = fields[8].toInt() != 0,
                source = fields[9],
                rssi = fields[10].toInt(),
            )
        }.getOrNull()
    }
}

data class StatusPacket(
    val session: String,
    val acknowledgedSequence: Long,
    val echoTimestampMs: Long,
    val commandedSteering: Int,
    val actualServoAngle: Int,
    val commandedThrottle: Int,
    val actualThrottle: Int,
    val failsafe: Boolean,
    val source: String,
    val rssi: Int,
) {
    val latencyMs: Long
        get() = (System.currentTimeMillis() - echoTimestampMs).coerceAtLeast(0L)
}

File: robot/jetson_nano/mcqueen_edge/protocol.py
"""Phone <-> McQueen Jetson UDP protocol.

Kept compatible with the existing Android controller.
"""

TOKEN = "rc-car"

PACKET_HELLO = "H"
PACKET_COMMAND = "C"
PACKET_EMERGENCY = "E"

MAX_STEERING = 1000
MAX_THROTTLE = 1000


def clamp(value, minimum, maximum):
    return max(minimum, min(maximum, value))


def parse_phone_packet(payload):
    """Parse one Android UDP packet.

    Returns a dictionary on success, or None for an invalid packet.
    """
    try:
        if isinstance(payload, bytes):
            text = payload.decode("ascii", errors="replace").strip()
        else:
            text = str(payload).strip()

        fields = [field.strip() for field in text.split(",")]

        if len(fields) < 5:
            return None

        packet_type = fields[0]

        if packet_type not in (
            PACKET_HELLO,
            PACKET_COMMAND,
            PACKET_EMERGENCY,
        ):
            return None

        if fields[1] != TOKEN:
            return None

        packet = {
            "type": packet_type,
            "session": fields[2],
            "sequence": int(fields[3]),
            "timestamp_ms": int(fields[4]),
        }

        if packet_type == PACKET_COMMAND:
            if len(fields) != 8:
                return None

            packet["steering"] = clamp(
                int(fields[5]), -MAX_STEERING, MAX_STEERING
            )
            packet["throttle"] = clamp(
                int(fields[6]), -MAX_THROTTLE, MAX_THROTTLE
            )
            packet["motor_enabled"] = int(fields[7]) != 0

        return packet

    except (ValueError, IndexError, TypeError):
        return None


def make_status(
    session,
    sequence,
    echo_timestamp_ms,
    commanded_steering,
    actual_servo_angle,
    commanded_throttle,
    actual_throttle,
    failsafe,
    source="JETSON",
    rssi=-1,
):
    """Generate the exact 11-field status packet Android expects."""
    return (
        "S,{},{},{},{},{},{},{},{},{},{}\n".format(
            session,
            int(sequence),
            int(echo_timestamp_ms),
            int(commanded_steering),
            int(actual_servo_angle),
            int(commanded_throttle),
            int(actual_throttle),
            1 if failsafe else 0,
            source,
            int(rssi),
        )
    )
